### DISCUSSION TITLE
How to perform colour correction with a "ColorChecker" reference image?

### DISCUSSION THREAD

**Discussion by bsdis** [colour-science/colour#747] [Q&A] [answered]
Hi, 
I have images of a colorchecker and I would like to perform color calibration of these images.
As I understand I have to do extract the 24 colour squares of the colorchecker, Convert from sRGB to RGB by applying inverse gamma, convert RGB to XYZ using 3x3 matrix, convert XYZ to Lab using standard formula and D6500 white point - perform a regression to find conversion coefficients - and then convert all pixels using the coefficients and the finally transform the converted image back into sRGB space.
Is that correctly understood, and is that possible to do using this library? (all the color conversions - not the extractions from images etc)
If yes - are there any examples available?

thanks

**Comment by thomasmansencal** [comment, marked as answer]
Hi @bsdis,

It is absolutely possible and I do that quite regularly.

Assuming you have the data for the colour rendition chart, at the basic minimum, this is what you need to do: https://github.com/colour-science/colour/blob/c3245b1830323991823f20580f81eee856dcc04c/colour/characterisation/correction.py#L852

We have a [repository](https://github.com/colour-science/colour-checker-detection) which does extraction and provides an example of colour correction here: https://github.com/colour-science/colour-checker-detection/blob/develop/colour_checker_detection/examples/examples_detection.ipynb

**Reply by tahir-aidesk** [reply]
@KelSolaar 

I can't find https://github.com/colour-science/colour-checker-detection/blob/develop/colour_checker_detection/examples/examples_detection.ipynb

I am trying to solve the same problem.

**Reply by larsmaxfield** [reply]
@tahir-aidesk At [commit d00540a](https://github.com/colour-science/colour-checker-detection/commit/d00540a2bff676e5e0fbc10313d1671ccba71db7) I found the now deleted color correction example:

[https://github.com/colour-science/colour-checker-detection/blob/d00540a2bff676e5e0fbc10313d1671ccba71db7/colour_checker_detection/examples/examples_detection.ipynb](https://github.com/colour-science/colour-checker-detection/blob/d00540a2bff676e5e0fbc10313d1671ccba71db7/colour_checker_detection/examples/examples_detection.ipynb).

You may need to use "Download raw file" if it doesn't show:

![image](https://github.com/user-attachments/assets/fce5635e-3157-4263-8c8a-6a7606ba0ab4)

**Reply by tahir-aidesk** [reply]
Thanks.

**Comment by bsdis** [comment]
Hi @KelSolaar 
Ahh ok. Really nice :)
I hope its ok to ask some follow up questions:
1) In the notebook you preprocess your images with `dcraw -T -g 2.4 12.92 *.CR2`. This basically encodes images as sRGB png images right?
2) I ran your example on 4 test images i had of the same scene where i have changed the lighting in each scene.
I get this result which is pretty nice actually
<img width="772" alt="Screenshot 2020-06-09 at 09 57 07" src="https://user-images.githubusercontent.com/1256202/84121632-b5e79e00-aa37-11ea-8e87-e18b9edb61ee.png">

But it seems that especially if you look on the white background its still not the same on the "after calibration" images. Is this to be expected? Can that be fixed somehow?

**Comment by bsdis** [comment]
Here is actually a better image. I have multiplied with and capped at 255, which also shows that some channels are saturated after the calibration (First 4 are before calibration and last 4 are after).

<img width="621" alt="Screenshot 2020-06-09 at 10 23 00" src="https://user-images.githubusercontent.com/1256202/84124276-4a9fcb00-aa3b-11ea-9db7-0e7c05caab97.png">

Is this because i need to do a white balancing also?

**Comment by thomasmansencal** [comment]
Hi @bsdis,

This is a great example of why relying on a colour rendition chart for colour correction is not the solution to all the imaging calibration problems as `$COLOUR_RENDITION_CHART_MAKER` claims it is.

What is happening here is that your scene illumination is changing while the ColorChecker Classic values are given for a particular illuminant, i.e. CIE D50. Ideally for things to really work, your scene should be lit with CIE D50 so that you can perform colour correction against the published values.

Ideally, you would need to measure the spectral irradiance of the scene and the spectral reflectance of the ColorChecker Classic (with a spectro-photometer) and recompute its values for the particular scene.

Alternatively, something cheap is to neutralise on the bottom row after the colour correction but it is obviously not ideal!

I hope that makes sense!

**Comment by bsdis** [comment]
Hi @KelSolaar 

Aha - yes that makes perfect sense. I have to keep that in mind.
If I first wanted to try the cheap method, you say "neutralise on the bottom row after the colour correction". How exactly is that done? In your code, the final image is generated by doing:
```
    im = colour.cctf_encoding(
        colour.colour_correction(
            COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES))
```
So I guess `im` is in sRGB colorspace? Would i do the neutralise directly on this image, or would I convert it to another colorspace first?
Also - would you cut a square from each of the bottom patches to neutralize after, and how would you do it?

Sorry - I hope its not to many questions.

Thanks :)

**Comment by thomasmansencal** [comment]
Yeah, so you would need to neutralise before that conversion, colour correction, white-balancing  and, most of the operations must ideally be done on floating-point linear data. The idea is to compute the average RGB colour of one of the swatches of the bottom row, we commonly take the third from the bottom right in the VFX industry, divide your image by it so that this swatch is white (or close to it because you only took the average) and then multiply the image by the value of that swatch, it should be around [0.19~, 0.19~, 0.19~] for sRGB.

**Comment by bsdis** [comment]
Hi @KelSolaar 

Hmm I think I might have misunderstood you.
I understood it like adding these 2 lines:
```
    im_ = COLOUR_CHECKER_IMAGES[i]/colour_checker.data['neutral 5 (.70 D)']
    im *= im_
```
So code becomes like this
```
for i, swatches in enumerate(SWATCHES):
    swatches_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches, D65, D65, colour.RGB_COLOURSPACES['sRGB'].RGB_to_XYZ_matrix))

    colour_checker = colour.characterisation.ColourChecker(
        os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i]),
        OrderedDict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_xyY)),
        D65)
    
    plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])
    
    swatches_f = colour.colour_correction(swatches, swatches, REFERENCE_SWATCHES)
    swatches_f_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches_f, D65, D65, colour.RGB_COLOURSPACES['sRGB'].RGB_to_XYZ_matrix))
    colour_checker = colour.characterisation.ColourChecker(
        '{0} - CC'.format(os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i])),
        OrderedDict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_f_xyY)),
        D65)
    
    plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])
    
    ##### ADDED WHITE BALANCING HERE:
    im_ = COLOUR_CHECKER_IMAGES[i]/colour_checker.data['neutral 5 (.70 D)'] # divide using broadcasting
    im *= im_
    ##########################
    im = colour.cctf_encoding(
        colour.colour_correction(
            im, swatches, REFERENCE_SWATCHES))
    b4after.append( [COLOUR_CHECKER_IMAGES[i],im] )
    plot_image(im);
```

But this gives me quite strange results. 
What did i misunderstand here?

**Comment by thomasmansencal** [comment]
You would need to divide by the average value of the colour checker in the image, i.e. `COLOUR_CHECKER_IMAGES[I]` and multiply by that of the reference one colour_checker.data['neutral 5 (.70 D)']

**Comment by bsdis** [comment]
As I understand the code, the following variables are:

```COLOUR_CHECKER_IMAGES[i] <--- Sample image```

```colour_checker.data['neutral 5 (.70 D)']) <--- average value of neutral 5 swatch of sample image```

```REFERENCE_COLOUR_CHECKER.data['neutral 5 (.70 D)'] <--- reference value of colorchecker for neutral 5```

So: divide by the average value of the colour checker in the image, i.e. COLOUR_CHECKER_IMAGES[I] and multiply by that of the reference one 
would be:
```
im = (COLOUR_CHECKER_IMAGES[i]/colour_checker.data['neutral 5 (.70 D)'])*\
        REFERENCE_COLOUR_CHECKER.data['neutral 5 (.70 D)']
```
Is that correct?

However it still gives quite different backgrounds though. 

<img width="264" alt="Screenshot 2020-06-09 at 22 38 34" src="https://user-images.githubusercontent.com/1256202/84200329-d995fd00-aaa6-11ea-9a84-12720c2c03d2.png">

So I guess thats due to my scene illumination changing while the ColorChecker Classic values are given for a particular illuminant like you mentioned, right?

What I actually wanted to achieve was to be able to take an arbitrary image under an arbitrary illumination and and arbitrary camera, and somehow calibrate those into same colorspace, so that I could directly compare the colors in the in images. In each picture there would always be a colorchecker of course. 
So I guess thats actually impossible then?

**Comment by thomasmansencal** [comment]
> Is that correct?

Not really, you would need to divide each given image by its own colour checker Neutral 5 swatch average, because each image has its own illuminant, you need to neutralise for each one of them accordingly and individually.

> So I guess thats due to my scene illumination changing while the ColorChecker Classic values are given for a particular illuminant like you mentioned, right?

Assuming you have neutralised each one of them, you will see that not only their whites are closer, but also that their exposure is roughly the same! You will still have discrepancies though because of the illuminant differences.

**Comment by bsdis** [comment]
Hi @KelSolaar 
Hope its okay i keep this thread running.
In your notebook (https://github.com/colour-science/colour-checker-detection/blob/develop/colour_checker_detection/examples/examples_detection.ipynb) 
you do:
```
    plot_image(colour.cctf_encoding(
        colour.colour_correction(
            COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES)));
```

the output of colour.cctf_encoding seem to be in the range -0.52 to 1.47. Should that not be in the range 0 to 1?

**Comment by thomasmansencal** [comment]
Hi,

> Hope its okay i keep this thread running.

Absolutely and good question! We never (pretty much never) clip data in Colour because this is what allows inversion of the transforms. Yes, you will often get negative values and values over one, this is by design. Whether it makes sense is something we are expecting the user to assess and is dependent on the usage. For your current case, because you are plotting an image, it would make sense to clip the values. They are effectively being clipped by `plot_image` just before being sent to `plt.imshow` because Matplotlib will raise an exception otherwise.

**Comment by bsdis** [comment]
I see. Instead of clipping, wouldn't it make more sense to keep the information as much as possible by shifting to min=0 and divide with max value to get 0<=x<=1 ?

**Comment by thomasmansencal** [comment]
Good question again!

We cannot really decide for the user here, given we are often working with negative out-of-gamut colours or very large HDR values, e.g. [1.e7, 1e9], the image would be entirely black except for a small speckle, i.e. the Sun, in a lot of cases.

A single large value is needed to throw you off trajectory here and it happens all the time in the type of applications we use Colour for. We really do swim in an ocean of **nans** and **infs** :)

**Comment by bsdis** [comment]
haha yes ok. thats a really good point :D
Btw. I tried changing method to method='Finlayson 2015', but that seems to give exactly the same results as if that parameter is omitted....even down to the 12th decimal...isn't that a bit curious? Maybe the code always redirects to the same method nomatter what the user sets?

**Comment by thomasmansencal** [comment]
depending on the number of terms/degree you use, you will get the same results, basically, all the methods are equivalent for degree 1 (linear).

**Comment by bsdis** [comment]
@KelSolaar I was hoping I could maybe get your opinion on the path I have chosen so far.

My usecase is that I need to compare colors in different images across many images. 

The challenge here is that I have no idea about the lighting used on each image and I have no idea about the camera used for each image (could be some phone cam or some DSLR cam or whatever). The only thing I know is that an xrite colorchecker is present on each image taken. Input images are in sRGB jpeg which is default output from many devices.. 
As Initially stated i need to be able to compare colors across all images, which means i need to transform all images into same colorspace-coordinate system so to say. I know that it will be impossible to perfectly normalize the colors as long as i have no control of the lightsource and the device, but I am looking to do it as good as possible. For future images i can change the xrite colorchecker with something more appropriate if need be. Is the `color-correction` course the right way to go for a problem like that?

**Comment by thomasmansencal** [comment]
I would maybe try to see if you can infer the illuminant from the scene, there is a great deal of literature on the topic, e.g. https://www.groundai.com/project/rehabilitating-the-color-checker-dataset-for-illuminant-estimation/. I never tried any of that unfortunately but it is certainly on my radar when time permits!

Otherwise, your approach is certainly correct!

**Comment by thomasmansencal** [comment]
This publication is of particular interest: https://arxiv.org/abs/1504.04548

**Comment by bsdis** [comment]
AHa...that sounds really interesting... I will look into that. Thanks a lot for the pointers! Will be sure to feed back what I find out :)

**Comment by thomasmansencal** [comment]
Closing this one for now! Feel free to add comments and I will re-open!

**Comment by alianoroozi** [comment]
Hi @KelSolaar,

I have some questions:

1. Should we apply white-balancing (neutralization) after color correction? If yes, should we use mean value of 'neutral 5 (.70 D)' swatch after color correction or before it? I am guessing after correction.
2. Should we do white-balancing in Lab color space or sRGB space?

Thanks

**Reply by thomasmansencal** [reply]
Hi,

From a conceptual standpoint, there are _only_ two places to white-balance:

-  In the camera sensitivities space, e.g. Camera RGB.
- In the Standard Human Observer Cone Responses (LMS) space.

Whether to choose one space over the other depends on the usage of the imagery. As to why _only_ those two spaces: those are the only places where there is theoretically no channel correlation or the least of it. Any other space might have channel cross-correlation and thus because the basis are not aligned with the camera or human observer responses basis will introduce some form of skewing in the colours.

Now to go back to your questions and starting with the last, i.e. 2, white balancing would ideally happen in cone space, i.e. using `colour.chromatic_adaptation` definition where you pick the source and target CIE XYZ tristimulus values. As for question 1, colour correction should white balance the source, within reason, if the target is white balanced. You could effectively neutralize after the fact, which would be mostly fine given that the colour correction matrix has introduced a lot of RGB channel correlation.

Here is something of interest we wrote a few months ago: https://www.colour-science.org/posts/the-colorchecker-considered-mostly-harmless/

**Reply by alianoroozi** [reply]
Thanks

**Comment by MMehdiMousavi** [comment]
Hi, Im trying the notebook to do a color correction of a custom image. However, I cant seem to get the thing to run correctly at all (even with no modifications). this is the error I encounter: 

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-f039509c97de> in <module>()
      2 for image in COLOUR_CHECKER_IMAGES:
      3     for swatches, colour_checker, masks in detect_colour_checkers_segmentation(
----> 4         image, additional_data=True):
      5         SWATCHES.append(swatches)
      6 

C:\Users\Mehdi\AppData\Roaming\Python\Python36\site-packages\colour_checker_detection\detection\segmentation.py in detect_colour_checkers_segmentation(image, samples, additional_data)
    767     colour_checkers_colours = []
    768     colour_checkers_data = []
--> 769     for colour_checker in extract_colour_checkers_segmentation(image):
    770         colour_checker = cctf_decoding(
    771             as_float_array(colour_checker[..., ::-1]) / 255)

C:\Users\Mehdi\AppData\Roaming\Python\Python36\site-packages\colour_checker_detection\detection\segmentation.py in extract_colour_checkers_segmentation(image)
    691     for colour_checker in colour_checkers_coordinates_segmentation(image):
    692         colour_checker = crop_and_level_image_with_rectangle(
--> 693             image, cv2.minAreaRect(colour_checker))
    694         width, height = (colour_checker.shape[1], colour_checker.shape[0])
    695 

C:\Users\Mehdi\AppData\Roaming\Python\Python36\site-packages\colour_checker_detection\detection\segmentation.py in crop_and_level_image_with_rectangle(image, rectangle)
    459     width_r, height_r = as_int_array([width_r, height_r])
    460 
--> 461     M_r = cv2.getRotationMatrix2D(centroid, angle, 1)
    462 
    463     image_r = cv2.warpAffine(image, M_r, (width, height), cv2.INTER_CUBIC)

TypeError: Can't parse 'center'. Sequence item with index 0 has a wrong type
```

anyone know how to fix this?

**Reply by thomasmansencal** [reply]
Hi @MMehdiMousavi,

The released version of "colour-checker-detection" has some know problems with latest versions of OpenCV. Please try the develop branch and let us know how it goes! We are working to release the new version soon.

Cheers,

Thomas

**Reply by MMehdiMousavi** [reply]
Hi and Thanks for the reply. 
I was able to make it work using opencv 4.4.0,
Color correction isn't really my forte so I appreciate it if you can point me in the right direction. 
I'm trying to color correct a [synthetic image](https://pasteboard.co/4zB72wX6zZFx.png), where the scene illumination is affected by heavy scattering.  Im including a virtual color chart as a reference so that I can find the correct color transformation filter (where the colors on the chart are exactly what they should be) and save it for reuse. 

from what I understand reading the discussion above, that depends on the lighting profile the chart was lit with, is there any other way to achieve results without manually messing with the colors?

Thanks

**Reply by thomasmansencal** [reply]
Hi,

This is a bit of a complex scenario here because you not only have to account for the illumination but also water absorption whose effect changes as a function of distance. I would recommend reading the [Sea-Thru publication](https://www.deryaakkaynak.com/sea-thru) as it gives a good idea of the type of process involved.

Cheers,

Thomas

**Comment by rakeshwar07** [comment]
Hi @KelSolaar @bsdis @alianoroozi ,

I have been following your conversation with bsdis. It was very usefull. Instead of 24 colour average I am using a algorithm to find the dominant colour. It seems to work. But I am consistently facing an issue with calibrating some images that has more brigtness in it.

Also, I tried the white balancing as you mentioned earlier in the threat. But it woresens my problem after including it. I am attaching the code and the results. Please let me know where I am going wrong. And what I can do to solve that type of issue
![Capture](https://user-images.githubusercontent.com/57210150/163794625-75233419-4aed-4dcb-bbbe-a1555c78ea10.JPG)
?
![image](https://user-images.githubusercontent.com/57210150/163794922-3c05b902-9161-4649-9f77-f967e0fdc348.png)

**Reply by thomasmansencal** [reply]
Your reference RGB colours here are expressed as 8-bit, most likely encoded values, they need to be represented as linear floating point values before attempting any correction.

**Reply by tahir-aidesk** [reply]
When I compare the color values of the ColorChecker chart to those provided by the Colour-Science API, I notice they don't match. I'm sure I'm missing something—could you please help me understand what might be going wrong?

https://xritephoto.com/documents/literature/en/ColorData-1p_EN.pdf here are xritephoto. 

and my response on `colorchecker = colour.CCS_COLOURCHECKERS['ColorChecker24 - After November 2014']`

is as follows 

ColorChecker24 - After November 2014 Values
------------------------------------------------------------
Patch: dark skin
XYZ: [0.4375, 0.3785, 0.0983]
RGB: [229, 147, 64]
Hex: #E59340  Swatch: [ #E59340 ]
------------------------------------------------------------
Patch: light skin
XYZ: [0.4222, 0.3725, 0.3362]
RGB: [208, 150, 150]
Hex: #D09696  Swatch: [ #D09696 ]
------------------------------------------------------------
Patch: blue sky
XYZ: [0.2761, 0.2984, 0.1786]
RGB: [159, 149, 106]
Hex: #9F956A  Swatch: [ #9F956A ]
------------------------------------------------------------
Patch: foliage
XYZ: [0.3733, 0.4512, 0.1347]
RGB: [179, 186, 75]
Hex: #B3BA4B  Swatch: [ #B3BA4B ]
------------------------------------------------------------
Patch: blue flower
XYZ: [0.3029, 0.2863, 0.2287]
RGB: [175, 138, 124]
Hex: #AF8A7C  Swatch: [ #AF8A7C ]
------------------------------------------------------------
Patch: bluish green
XYZ: [0.2864, 0.3897, 0.4144]
RGB: [98, 183, 165]
Hex: #62B7A5  Swatch: [ #62B7A5 ]
------------------------------------------------------------
Patch: orange
XYZ: [0.5281, 0.4053, 0.3126]
RGB: [247, 140, 144]
Hex: #F78C90  Swatch: [ #F78C90 ]
------------------------------------------------------------
Patch: purplish blue
XYZ: [0.2324, 0.2112, 0.1091]
RGB: [165, 116, 82]
Hex: #A57452  Swatch: [ #A57452 ]
------------------------------------------------------------
Patch: moderate red
XYZ: [0.5044, 0.3271, 0.1890]
RGB: [255, 102, 112]
Hex: #FF6670  Swatch: [ #FF6670 ]
------------------------------------------------------------
Patch: purple
XYZ: [0.3335, 0.2506, 0.0628]
RGB: [213, 108, 52]
Hex: #D56C34  Swatch: [ #D56C34 ]
------------------------------------------------------------
Patch: yellow green
XYZ: [0.3989, 0.5041, 0.4332]
RGB: [149, 200, 165]
Hex: #95C8A5  Swatch: [ #95C8A5 ]
------------------------------------------------------------
Patch: orange yellow
XYZ: [0.4936, 0.4443, 0.4293]
RGB: [218, 164, 168]
Hex: #DAA4A8  Swatch: [ #DAA4A8 ]
------------------------------------------------------------
Patch: blue
XYZ: [0.2052, 0.1687, 0.0560]
RGB: [165, 97, 53]
Hex: #A56135  Swatch: [ #A56135 ]
------------------------------------------------------------
Patch: green
XYZ: [0.3230, 0.5104, 0.2233]
RGB: [108, 211, 108]
Hex: #6CD36C  Swatch: [ #6CD36C ]
------------------------------------------------------------
Patch: red
XYZ: [0.5629, 0.3356, 0.1278]
RGB: [255, 84, 88]
Hex: #FF5458  Swatch: [ #FF5458 ]
------------------------------------------------------------
Patch: yellow
XYZ: [0.4678, 0.4760, 0.5993]
RGB: [185, 181, 198]
Hex: #B9B5C6  Swatch: [ #B9B5C6 ]
------------------------------------------------------------
Patch: magenta
XYZ: [0.4213, 0.2669, 0.1895]
RGB: [239, 89, 114]
Hex: #EF5972  Swatch: [ #EF5972 ]
------------------------------------------------------------
Patch: cyan
XYZ: [0.2091, 0.3027, 0.1806]
RGB: [98, 164, 105]
Hex: #62A469  Swatch: [ #62A469 ]
------------------------------------------------------------
Patch: white 9.5 (.05 D)
XYZ: [0.3489, 0.3642, 0.8807]
RGB: [102, 166, 241]
Hex: #66A6F1  Swatch: [ #66A6F1 ]
------------------------------------------------------------
Patch: neutral 8 (.23 D)
XYZ: [0.3456, 0.3599, 0.5900]
RGB: [142, 163, 199]
Hex: #8EA3C7  Swatch: [ #8EA3C7 ]
------------------------------------------------------------
Patch: neutral 6.5 (.44 D)
XYZ: [0.3441, 0.3592, 0.3649]
RGB: [166, 161, 156]
Hex: #A6A19C  Swatch: [ #A6A19C ]
------------------------------------------------------------
Patch: neutral 5 (.70 D)
XYZ: [0.3458, 0.3591, 0.1906]
RGB: [183, 159, 107]
Hex: #B79F6B  Swatch: [ #B79F6B ]
------------------------------------------------------------
Patch: neutral 3.5 (1.05 D)
XYZ: [0.3426, 0.3576, 0.0882]
RGB: [190, 158, 56]
Hex: #BE9E38  Swatch: [ #BE9E38 ]
------------------------------------------------------------
Patch: black 2 (1.5 D)
XYZ: [0.3438, 0.3561, 0.0315]
RGB: [196, 157, 0]
Hex: #C49D00  Swatch: [ #C49D00 ]
------------------------------------------------------------

Process finished with exit code 0

**Comment by danoneata** [comment]
Hello and thank you for the very nice and comprehensive library!

I'm trying to perform color correction in a scenario in which I have a strong shading (vignetting) effect and I would need some advice on how is best to proceed.

Concretely, the image of the checker looks like this.
![image](https://user-images.githubusercontent.com/819256/192279783-d47ccedd-082c-41d4-872c-37127677e866.png)

In order to remove the vignetting effect I divide the above image by an image of the white board, obtaining the following result:
![image](https://user-images.githubusercontent.com/819256/192280162-51537933-da54-496a-aecc-5db92a935346.png)

I have a couple of questions.

1. When converting the reference colors to (linear) RGB (`XYZ_to_RGB`) I need to provide the illuminant of the scene where the photo is taken. What value should I use if I work on the vignette-corrected image? Currently, I'm using `colour.RGB_COLOURSPACES["sRGB"].whitepoint` (which corresponds to the D65 illuminant, as far as I understand), but I'm not sure if this is correct:
```python
REFERENCE_SWATCHES = colour.XYZ_to_RGB(
    colour.xyY_to_XYZ(list(REFERENCE_COLOUR_CHECKER.data.values())),
    REFERENCE_COLOUR_CHECKER.illuminant,
    colour.RGB_COLOURSPACES["sRGB"].whitepoint,
    colour.RGB_COLOURSPACES["sRGB"].matrix_XYZ_to_RGB,
)
```
(I know that the light used for capturing the original image had a temperature of 5000K, but I don't know it's spectrum.)

2. It looks like the results are sensitive to the location of the checker in the original image the checker. In some cases the white background gets a yellowish tint:
- original image
![image](https://user-images.githubusercontent.com/819256/192283261-b67f92f6-199c-4b22-ad72-99ca70d02a08.png)
- after vignette correction
![image](https://user-images.githubusercontent.com/819256/192282538-505941fd-978f-4bf2-8e97-b8580e4d3903.png)
- after color correction 
![image](https://user-images.githubusercontent.com/819256/192282577-023cb5d7-ac53-4bed-a001-aaf4ad38d00e.png)
(For the above experiment, I've used a linear function—Finlayson 2015, with degree 1 and no root expansion.)

Thank you in advance and forgive my lack of knowledge; I was not familiar with intricacies of color science until very recently!

**Comment by thomasmansencal** [comment]
Hello @danoneata,

1. It depends what on what you are trying to achieve. Typically, you would use your working space illuminant, e.g. D65 if sRGB.
2. This is expected as there is no colour preference in the fitting. Put another way, the error is minimised without favouring a colour over another thus achromatic colours can become chromatic. You can neutralise the output as a last step.

Cheers,

Thomas

**Reply by danoneata** [reply]
Thanks a lot for the reply, Thomas!

1. I'm a bit confused. Where does the illuminant used to light the scene come into play in the color correction process? From [your very persuasive post](https://www.colour-science.org/posts/the-colorchecker-considered-mostly-harmless/) I understand that illumination is an important aspect in getting sensible color correction results, but in [the example](https://github.com/colour-science/colour-checker-detection/blob/develop/colour_checker_detection/examples/examples_detection.ipynb) I don't know where this information is used. Initially, I thought that one of parameters passed to the `colour.XYZ_to_RGB` correspond to the scene illuminant, but now I'm not sure about this. For example, if my scene was lit by a light with temperature 5000K, wouldn't using D50 be more appropriate than D65?
2. Thanks! I think balancing one of the neutral tones is a good idea. One thing that I'm observing, is that while I fix one of the values (e.g., white), the others may still preserve some chromacity. But I don't know if there is a fix for this; or maybe is pointing to some other flaw in my setup.

**Reply by thomasmansencal** [reply]
The illuminant is critical when you are in a position where you can measure spectral data: At this point and with knowledge of the spectral reflectances of the chart and that of the scene illumination, you can generate precise reference values. Without the spectral data changing the whitepoint will be very much akin to white-balancing, so it is not important. You might as well pick the working space one because having white staying white is probably the next important aspect.

**Reply by danoneata** [reply]
Thanks a lot for all the explanations, Thomas! They were very instructive!

Just a quick update: the reason why I had trouble fitting the neutrals was simply because the original image was already in the linear space and I was unaware of this, so I'd applied a non-linear decoding function 😐
